crypto/jwks/component.go

```diff
@@ -159,21 +173,30 @@ func (k *jwksCrypto) initJWKSFromURL(ctx context.Context, url string) error {
 	return nil
 }
 
-func (k *jwksCrypto) initJWKSFromFile(ctx context.Context, file string) error {
+func (k *jwksCrypto) initJWKSFromFile(parentCtx context.Context, file string) error {
+	ctx, cancel := context.WithCancel(parentCtx)
+
 	// Get the path to the folder containing the file
 	path := filepath.Dir(file)
 
 	// Start watching for changes in the filesystem
 	eventCh := make(chan struct{})
-	loaded := make(chan error, 1) // Needs to be buffered to prevent an (unlikely, but possible) goroutine leak
+	loadedCh := make(chan error, 1) // Needs to be buffered to prevent an (unlikely, but possible) goroutine leak
+
+	k.wg.Add(2)
+
 	go func() {
-		watchErr := fswatcher.Watch(k.ctx, path, eventCh)
+		defer k.wg.Done()
+		watchErr := fswatcher.Watch(ctx, path, eventCh)
 		if watchErr != nil && !errors.Is(watchErr, context.Canceled) {
 			// Log errors only
```

Comment:
It feels like we should be failing the component when there is an error here.

Reply:
Perhaps, but what if the error happens after the component has started?

Reply:
The component should be closed imo.

Reply:
I don't think we should cause a panic in the entire runtime for this. And closing the component would mean that the app would stop working until it's manually restarted.

When this fails btw the component can continue to work with the keys that have been loaded in memory, it just can't pick up new keys.

Reply:
I'm not suggesting that the entire runtime to panic, just that this component to not be in a ready state and usable. The rest of the dapr runtime that does not rely on this should continue to function.

> When this fails btw the component can continue to work with the keys that have been loaded in memory, it just can't pick up new keys.

I am concerned that this is going to lead to silent failure modes and cause larger disruption downstream which are difficult to debug.

Reply:
> I'm not suggesting that the entire runtime to panic, just that this component to not be in a ready state and usable. The rest of the dapr runtime that does not rely on this should continue to function.

Wouldn't that cause silent failures too?

Reply:
> > I'm not suggesting that the entire runtime to panic, just that this component to not be in a ready state and usable. The rest of the dapr runtime that does not rely on this should continue to function.
> 
> Wouldn't that cause silent failures too?

We should have a mechanism for reporting health on a per component basis. This can be some combination of kube resource status, logs, metrics, error codes. Only printing an error log line is very difficult for a human to pick up, and impossible for a machine to self heal automatically.